### PR TITLE
[SSP-2929] Allow to move internal_sequence up and down

### PR DESCRIPTION
**Pull request by bzwei**
https://issues.redhat.com/browse/SSP-2929

Add endpoint post `/workflows/{id}/reposition` to adjust the internal_sequence. Added backend support.

### CODE REVIEW

**Comment by bzwei** [comment]
@cutwater please help with the formatting error:
flake8 requires no space before `:`
```
         return query.values_list("internal_sequence", flat=True)[
             (delta - 1): (delta + 1)
         ]
```
tox -eblack requires space before and after `:`
```
         return query.values_list("internal_sequence", flat=True)[
             (delta - 1) : (delta + 1)
         ]
```

**Comment by cutwater** [comment]
@bzwei I updated flake8 configuration #542. When using `black`, the `E203` error can be safely ignored, because whitespaces around semicolons are handle by black and in this particular case it is false positive flake8 error.

**Comment by bzwei** [comment]
@cutwater Still got E203 error.

File: pinakes/main/approval/models.py [hunk 1]
```diff
@@ -187,10 +192,96 @@ class Meta:
...
+            .values_list("internal_sequence", flat=True)
+            .first()
+        )
+        return first_sequence / 2
+
     def __str__(self):
         return self.name
 
 
+@receiver(pre_save, sender=Workflow)
```
[the remark this answers is not in the document]
Comment (bzwei) [reply pinakes/main/approval/models.py:271]: It was moved from https://github.com/ansible/pinakes/blob/devel/pinakes/main/approval/views.py#L317.  `internal_sequence` is an internal field, we need logic to determine the value when a new object is created. It used to be in the view, I think it is more logic to move to model before the object is saved.

File: pinakes/main/approval/models.py [hunk 2]
```diff
@@ -184,10 +189,96 @@ class Meta:
             ),
         ]
 
+    def move_internal_sequence(self, delta):
```
Comment (cutwater) [inline pinakes/main/approval/models.py:192]: Can you please add some explanation of how this works and how do you use Decimal field for `Workflow.internal_sequence`.

File: pinakes/main/approval/models.py [hunk 3]
```diff
@@ -184,10 +189,96 @@ class Meta:
...
+
+        if delta > 0:
+            self.internal_sequence = Decimal(
+                self._internal_sequence_plus(delta)
+            )
+        else:
+            self.internal_sequence = Decimal(
+                self._internal_sequence_minus(-delta)
+            )
+        self.save()
```
Comment (cutwater) [inline pinakes/main/approval/models.py:208]: It's probably better to avoid implicit save of a model in a method that modifies it.

**Review by cutwater — CHANGES_REQUESTED** [review]

File: pinakes/main/approval/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bzwei) [reply pinakes/main/approval/models.py:275]: what's your recommendation if I want to auto assign the value before save?

**Review by cutwater — CHANGES_REQUESTED** [review]
If you described algorithm here it could've spared some time reading and understanding the logic behind this.

Unfortunately this algorithm is error prone and can be broken in 5 steps (10 in actual production environment).

Here is how you can reproduce it:

1. Create 3 workflows (e.g. `workflow-0`, `workflow-1` and `woroflow-2` with `internal_sequence` values 1, 2 and 3 respectively)
2. Move workflow-0 forward (`delta=1`)
3. Move workflow-1 forward (`delta=2`)
4. Repeat steps 2 and 3 until it breaks.

As a result eventually the `internal_sequence` reaches the precision of a decimal value stored in the database producing either, depending on whether the records were reloaded from the database between iterations.
* Incorrect order (in the particular example the `workflow-1` is moved ahead of `workflow-2`).
* Unique constraint violation and incorrect order.

Example:
```
step, workflow_order, internal_sequence
0     {workflow0,workflow1,workflow2} | {1.000000,2.000000,3.000000}
1     {workflow0,workflow1,workflow2} | {2.500000,2.750000,3.000000}
2     {workflow0,workflow1,workflow2} | {2.875000,2.937500,3.000000}
3     {workflow0,workflow1,workflow2} | {2.875000,2.937500,3.000000}
4     {workflow0,workflow1,workflow2} | {2.992188,2.996094,3.000000}
5     {workflow0,workflow2,workflow1} | {2.994141,3.000000,4.000000}
```

There are two key problems with this implementations:

1. Decimal in PostgreSQL and in Python have different precision. This explains the difference in behavior when records were reloaded from the database between iterations and when didn't. In Python decimal has default precision 28 places), Workflow model sets server side precision as only 6 places.
2. Because the precision is limited, the algorithm accumulates fraction until it goes out of boundaries.

**Comment by bzwei** [comment]
@cutwater First of all the sequence of workflows is not critical in execution. In production it is unlikely the user will have many workflows. The user will even more rarely move the same workflow recursively to break the order as illustrated. To prevent deterioration of the decimal numbers in a long run we can have a scheduled job to reset the sequence numbers to integers.

**Comment by eclarizio** [comment]
> @cutwater First of all the sequence of workflows is not critical in execution. In production it is unlikely the user will have many workflows. The user will even more rarely move the same workflow recursively to break the order as illustrated. To prevent deterioration of the decimal numbers in a long run we can have a scheduled job to reset the sequence numbers to integers.

@bzwei I think your point about not having many workflows is accurate, however I could still see scenarios in which 3 workflows exist and over time they could be reordered more than enough times to cause this issue. A scheduled job is a solution to the problem but that feels more like a workaround and that the issue should be addressed at the implementation level so that we don't even need to worry about having a separate task to "fix" the numbers. Since we have the opportunity to address it before it becomes a problem I think we should do it now, if we can. I'm not sure what the best solution is, though. Increasing the floating point precision would certainly give us more space, but it feels like we're still ignoring the core issue.

@cutwater @bzwei Without changing the API payloads and structure of how we are expecting the UI to call the reposition endpoints, is there a way we can think of doing this that avoids the issue? Or, should we abandon the current API signature and change it to a different strategy? I could see the UI having an "edit sequence" screen where the user can drag/drop as they see fit, but no calls get sent until they hit "save" and then it simply sends an array with the IDs of the objects in the sequence the user wants. This obviously changes the underlying model structure a bit, but it was the first thing I could think of.

**Comment by cutwater** [comment]
> Without changing the API payloads and structure of how we are expecting the UI to call the reposition endpoints, is there a way we can think of doing this that avoids the issue?

@eclarizio @bzwei Regarding current issue, I don't think the API change is necessary. I think it can be sufficient to fix the problems indicated above:

1. Resolve inaccuracy in calculations by using same precision in Python and PostgreSQL. It seems much easier to use integers instead of decimals, to avoid precision issues, but as far I'm aware catalog in cloud already considered that design and had issues with that. Could you please recall what the particular issue it was?
2. In case when the new position is in conflict with the existing record, we need to re-balance the affected records.

>  I could see the UI having an "edit sequence" screen where the user can drag/drop as they see fit, but no calls get sent until they hit "save" and then it simply sends an array with the IDs of the objects in the sequence the user wants.

It sounds like a nice improving UX feature, however it may require way more effort to implement it properly, as it brings new challenges in handling concurrency issues.

File: pinakes/main/approval/models.py [hunk 4]
```diff
@@ -184,9 +185,124 @@ class Meta:
...
+        if delta == 0:
+            return
+
+        if delta > 0:
+            self.internal_sequence = self._internal_sequence_plus(delta)
+        else:
+            self.internal_sequence = self._internal_sequence_minus(-delta)
+
+    def _internal_sequence_plus(self, delta):
+        if delta == math.inf:
```
Comment (cutwater) [inline pinakes/main/approval/models.py:215]: Usage `math.inf` \ `-math.inf` constants for indication of a first \ last positions is obcure.
For example you could use enum or constant parameter for a direction indication (e.g. `FIRST`, `LAST`, `FORWARD`, `BACKWARD`)

**Review by cutwater — CHANGES_REQUESTED** [review]

File: pinakes/main/approval/models.py [hunk 5]
```diff
@@ -184,9 +185,124 @@ class Meta:
...
+        return int(first_sequence / 2)
+
+    def _rebalance_internal_sequences(self):
+        workflows = Workflow.objects.filter(tenant=self.tenant)
+        new_sequence = 0
+        for workflow in workflows:
+            new_sequence = new_sequence + Workflow.INTERNAL_INTERVAL
+            workflow.internal_sequence = new_sequence
+
+        Workflow.objects.bulk_update(workflows, ["internal_sequence"], 100)
```
[the remark this answers is not in the document]
Comment (bzwei) [reply pinakes/main/approval/models.py:287]: This hardcoded number plays trivial factor. The value does not matter much and almost never needs adjustment.

File: pinakes/main/approval/models.py [hunk 6]
```diff
@@ -184,9 +185,124 @@ class Meta:
...
+            Workflow.objects.filter(tenant=self.tenant)
+            .values_list("internal_sequence", flat=True)
+            .first()
+        )
+        return self._half_first_internal_sequence(first_sequence)
+
+    def _half_first_internal_sequence(self, first_sequence):
+        if first_sequence < 2:
+            self._rebalance_internal_sequences()
+            return int(Workflow.INTERNAL_INTERVAL / 2)
```
[the remark this answers is not in the document]
Comment (bzwei) [reply pinakes/main/approval/models.py]: The number 2 is for average. It is part of the algorithm. A constant makes it harder to read.

File: pinakes/main/approval/models.py [hunk 7]
```diff
@@ -184,9 +185,124 @@ class Meta:
...
+            return self._internal_sequence_extend_from_last()
+
+        if len(sequence_range) == 1:
+            return sequence_range[0] + Workflow.INTERNAL_INTERVAL
+
+        if sequence_range[1] - sequence_range[0] < 2:
+            self._rebalance_internal_sequences()
+            return self._internal_sequence_plus(delta)
+
+        return int((sequence_range[0] + sequence_range[1]) / 2)
```
[the remark this answers is not in the document]
Comment (bzwei) [reply pinakes/main/approval/models.py]: Updated. Thanks.

File: pinakes/main/approval/models.py [hunk 8]
```diff
@@ -184,9 +185,124 @@ class Meta:
...
+            .values_list("internal_sequence", flat=True)
+            .first()
+        )
+        return self._half_first_internal_sequence(first_sequence)
+
+    def _half_first_internal_sequence(self, first_sequence):
+        if first_sequence < 2:
+            self._rebalance_internal_sequences()
+            return int(Workflow.INTERNAL_INTERVAL / 2)
+        return int(first_sequence / 2)
```
[the remark this answers is not in the document]
Comment (bzwei) [reply pinakes/main/approval/models.py]: it is the algorithm.

File: pinakes/main/approval/models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cutwater) [reply pinakes/main/approval/models.py]: I was referring to integer dividing here, not a constant. Apologies if it was misleading.

File: pinakes/main/approval/models.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (cutwater) [reply pinakes/main/approval/models.py]: Same here. Integer division.

File: pinakes/main/approval/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cutwater) [reply pinakes/main/approval/models.py:287]: This trivial factor doesn't seem to be local to the current query. Bulk updates may be used in other parts of the project. So hard coding it here would make less straightforward changing it in future if necessary.

File: pinakes/main/approval/models.py [hunk 9]
```diff
@@ -184,9 +184,123 @@ class Meta:
...
+        if delta == 0:
+            return
+
+        if delta > 0:
+            self.internal_sequence = self._internal_sequence_plus(delta)
+        else:
+            self.internal_sequence = self._internal_sequence_minus(-delta)
+
+    def _internal_sequence_plus(self, delta):
+        if delta > Workflow.objects.count():
```
Comment (cutwater) [inline pinakes/main/approval/models.py:213]: Can you please explain the change. 

In previous iterations delta was compared against `math.inf` \ `-math.inf` constants to determine if the record is moved to the last \ first position. Now it is an extra database query and comparison against total count of records in the database.

File: pinakes/main/approval/views.py [hunk 10]
```diff
@@ -313,22 +313,38 @@ def unlink(self, request, pk):
...
+                status=status.HTTP_400_BAD_REQUEST,
             )
-        serializer.save(
-            internal_sequence=next_seq,
-            tenant=Tenant.current(),
-        )
+
+        if "placement" in serializer.validated_data:
+            if serializer.validated_data["placement"] == "top":
+                offset = -math.inf
```
Comment (cutwater) [inline pinakes/main/approval/views.py:334]: Constants `math.inf` are still used to determine first \ last position here.

**Review by cutwater — CHANGES_REQUESTED** [review]

File: pinakes/main/approval/models.py (hunk 9 above)
Comment (bzwei) [reply pinakes/main/approval/models.py:213]: `delta` is expected to be a number to direct the change. When delta is big (out of bound) we can do something special because we know the new sequence will be definitely at the bottom or top. The caller does not need to use a constant to indicate it is bottom or top. For example delta = 100,000 will be effectively the same as using a constant `bottom`. The caller can however pass in `math.inf` to make sure delta is big enough.

This is an internal method, we try to normalize the argument. At the API level we do use constant `bottom` and `top` so the customer is not forced to use math.inf.

File: pinakes/main/approval/views.py (hunk 10 above)
Comment (bzwei) [reply pinakes/main/approval/views.py:334]: See my comments above. We don't need a constant, just need some big number.

File: pinakes/main/approval/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bzwei) [reply pinakes/main/approval/models.py:287]: https://docs.djangoproject.com/en/4.0/ref/models/querysets/ does not indicate the batch_size argument is global. I believe it is local. The default value is None, meaning all qualified workflows will be updated in one shot. We can use default value as well because in reality we don't expect the customer will add a lot of workflows. I choose 100 because I saw examples using this value.

File: pinakes/main/approval/models.py (hunk 9 above)
Comment (cutwater) [reply pinakes/main/approval/models.py:213]: The `math.inf` constant has a floating point type. Methods in the PR are based on integer types. The floating point type is not used anywhere else except these magic constants. This make the code unclear, hard to follow. Usage of inconsistent data types without a good reason makes this code smell.

File: pinakes/main/approval/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cutwater) [reply pinakes/main/approval/models.py:287]: It seems you understood my comment in a different way. The `100` number is not exclusive to that code. In case any other bulk_update operations appear, more hard coded numbers appear as a result. In case this number is considered sane default for the batch size, changing it across the code would require additional refactoring and searching for those magic numbers.

File: pinakes/main/approval/models.py (hunk 9 above)
Comment (cutwater) [reply pinakes/main/approval/models.py:213]: > delta is expected to be a number to direct the change. When delta is big (out of bound) we can do something special because we know the new sequence will be definitely at the bottom or top.

Please help me to understand the following. If you're comparing the `delta` with Workflow objects count, how does this code handle the situation when let's say the workflow is second to last and the delta is lower then total count.

For example:

Total workflows: 10
Current workflow has the position 9 out of 10.
The `delta` is 5.

File: pinakes/main/approval/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eclarizio) [reply pinakes/main/approval/models.py:287]: I think I agree with @cutwater here, the `100` seems like a fine default. I know it's currently only used in this one place, but it does seem to make sense to me to change it to a constant that is maybe more descriptive and easier to change later. Something like `DEFAULT_BULK_UPDATE_VALUE`?

File: pinakes/main/approval/models.py (hunk 4 above)
Comment (eclarizio) [reply pinakes/main/approval/models.py]: I like this idea, I think it would be easier to digest when reading through. I'm still not sure if `math.inf` and `-math.inf` are the "correct" values for the constant since they aren't integers, but it's much easier to understand that you're trying to compare against FIRST (or TOP).

**Review by cutwater — APPROVED** [review]
@bzwei Thank you for addressing my comments. You found a way how to eliminate floating point constants usage, which simplified the logic behind moving the internal sequence number. Good job!
